File: sample/sample-java/SampleFilterTablesLog.java
//---------------------------------------------------------------------------
//
// TSDuck sample Java application running a chain of plugins:
// Filter tables using --log-hexa-line to get binary tables in a Java class.
//
// See SampleFilterTablesEvent.java for an equivalent example using plugin
// events to get a binary content of the tables.
//
//----------------------------------------------------------------------------

import io.tsduck.AbstractAsyncReport;
import io.tsduck.Report;
import io.tsduck.TSProcessor;

public class SampleFilterTablesLog {

    /**
     * This string is a user-defined marker to locate the hexa line in the log.
     */
    private static final String LOG_PREFIX = "#TABLE#";

    /**
     * A pure Java class which handles TSDuck log messages.
     */
    private static class Logger extends AbstractAsyncReport {

        /**
         * Constructor.
         * @param severity Initial severity.
         */
        public Logger(int severity) {
            super(severity, false, 512);
        }

        /**
         * This method is invoked each time a message is logged by TSDuck.
         * @param severity Severity of the message.
         * @param message Message line.
         */
        @Override
        public void logMessageHandler(int severity, String message) {
            // Filter lines containing the marker.
            final int pos = message.lastIndexOf(LOG_PREFIX);
            if (pos >= 0) {
                // Got a hexadecimal representation of the table.
                String hexa = message.substring(pos + LOG_PREFIX.length());
                System.out.println("Table: " + hexa);
            }
            else {
                // This looks like a real log message.
                System.err.println(message);
            }
        }
    }

    /**
     * Main program.
     * @param args Command line arguments.
     */
    public static void main(String[] args) {

        // Create a user-defined report to log multi-threaded messages.
        Logger rep = new Logger(Report.Info);

        // Create a TS processor using the report.
        TSProcessor tsp = new TSProcessor(rep);

        // Set the plugin chain.
        tsp.input = new String[] {"http", "https://github.com/tsduck/tsduck-test/raw/master/input/test-001.ts"};
        tsp.plugins = new String[][] {{"tables", "--pid", "0", "--log-hexa-line=" + LOG_PREFIX}};
        tsp.output = new String[] {"drop"};

        // Run the TS processing and wait until completion.
        tsp.start();
        tsp.waitForTermination();
        tsp.delete();

        // Terminate the asynchronous report.
        rep.terminate();
        rep.delete();
    }
}

File: sample/sample-python/sample-filter-tables-log.py
#!/usr/bin/env python
#----------------------------------------------------------------------------
#
# TSDuck sample Python application running a chain of plugins:
# Filter tables using --log-hexa-line to get binary tables in a Python class.
#
# See sample-filter-tables-event.py for an equivalent example using plugin
# events to get a binary content of the tables.
#
#----------------------------------------------------------------------------

import tsduck, sys

# This string is a user-defined marker to locate the hexa line in the log.
LOG_PREFIX = "#TABLE#"

# A Python class which handles TSDuck log messages.
class Logger(tsduck.AbstractAsyncReport):
    # This method is invoked each time a message is logged by TSDuck.
    def log(self, severity, message):
        # Filter, locate, extract and parse the hexa output from plugin "tables".
        pos = message.find(LOG_PREFIX)
        if pos >= 0:
            # This is a log line containing an hexa dump.
            hexa = message[pos+len(LOG_PREFIX):]
            print("Table: %s" % (hexa))
        else:
            # This is a real log message, just display it.
            print(tsduck.Report.header(severity) + message, file=sys.stderr)

# Create an asynchronous report to log multi-threaded messages.
rep = Logger()

# Create a TS processor, set plugin chain.
tsp = tsduck.TSProcessor(rep)
tsp.input = ['http', 'https://github.com/tsduck/tsduck-test/raw/master/input/test-001.ts']
tsp.plugins = [ ['tables', '--pid', '0', '--log-hexa-line=' + LOG_PREFIX] ]
tsp.output = ['drop']

# Run the TS processing and wait until completion.
tsp.start()
tsp.waitForTermination()
tsp.delete()

# Terminate the asynchronous report.
rep.terminate()
rep.delete()
